crates/consensus/auto-seal/tests/it/auto_mine.rs

```diff
@@ -0,0 +1,125 @@
+//! auto-mine consensus integration test
+use clap::Parser;
+use jsonrpsee::{core::client::ClientT, http_client::HttpClientBuilder, rpc_params};
+use reth::{
+    cli::{
+        components::RethNodeComponents,
+        ext::{NoArgs, NoArgsCliExt, RethNodeCommandConfig},
+    },
+    node::NodeCommand,
+    runner::CliRunner,
+    tasks::TaskSpawner,
+};
+use reth_primitives::{hex, revm_primitives::FixedBytes, ChainSpec, Genesis};
+use reth_provider::CanonStateSubscriptions;
+use reth_transaction_pool::TransactionPool;
+use std::{sync::Arc, time::Duration};
+
+#[derive(Debug)]
+struct AutoMineConfig;
+
+impl RethNodeCommandConfig for AutoMineConfig {
+    fn on_node_started<Reth: RethNodeComponents>(&mut self, components: &Reth) -> eyre::Result<()> {
+        let pool = components.pool();
+        let mut canon_events = components.events().subscribe_to_canonical_state();
+
+        components.task_executor().spawn_critical_blocking("rpc request", Box::pin(async move {
+            // submit tx through rpc
+            let raw_tx = "0x02f876820a28808477359400847735940082520894ab0840c0e43688012c1adb0f5e3fc665188f83d28a029d394a5d630544000080c080a0a044076b7e67b5deecc63f61a8d7913fab86ca365b344b5759d1fe3563b4c39ea019eab979dd000da04dfc72bb0377c092d30fd9e1cab5ae487de49586cc8b0090";
+            let client = HttpClientBuilder::default().build("http://127.0.0.1:8545").expect("http client should bind to default rpc port");
+            let response: String = client.request("eth_sendRawTransaction", rpc_params![raw_tx]).await.expect("client request should be valid");
+            let expected = "0xb1c6512f4fc202c04355fbda66755e0e344b152e633010e8fd75ecec09b63398";
+
+            assert_eq!(&response, expected);
+
+            // more than enough time for the next block
+            let sleep = tokio::time::sleep(Duration::from_secs(15));
+            tokio::pin!(sleep);
+
+            // wait for canon event or timeout
+            tokio::select! {
+                _ = &mut sleep => {
+                    panic!("Canon state update took too long to arrive")
+                }
+
+                update = canon_events.recv() => {
+                    let event = update.expect("canon events stream is still open");
+                    let new_tip = event.tip();
+                    let expected_tx_root: FixedBytes<32> = hex!("c79b5383458e63fb20c6a49d9ec7917195a59003a2af4b28a01d7c6fbbcd7e35").into();
+                    assert_eq!(new_tip.transactions_root, expected_tx_root);
+                    assert_eq!(new_tip.number, 1);
+                    assert!(pool.pending_transactions().is_empty());
+                }
+            }
```

Comment:
should be able to use tokio timeout here: https://docs.rs/tokio/latest/tokio/time/fn.timeout.html

Reply:
good call